sysdump/encryption.go

```diff
@@ -0,0 +1,83 @@
+// Copyright 2021 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sysdump
+
+import (
+	"bytes"
+	_ "embed"
+	"io"
+	"os"
+	"path/filepath"
+	"time"
+
+	"golang.org/x/crypto/openpgp"
+	"golang.org/x/crypto/openpgp/packet"
+)
+
+//go:embed pubkey.gpg
+var pubKey []byte
+
+func createEncryptedZipFile(pathToZipFile string, pathToEncryptionKey string) (string, error) {
+	// Read the destination entity.
+	e, err := readDestinationEntity(pathToEncryptionKey)
+	if err != nil {
+		return "", err
+	}
+	// Create the destination file.
+	d := pathToZipFile + ".gpg"
+	o, err := os.Create(d)
+	if err != nil {
+		return "", err
+	}
+	// Encrypt the zip file.
+	h := &openpgp.FileHints{
+		FileName: filepath.Base(pathToZipFile),
+		IsBinary: true,
+		ModTime:  time.Now(),
+	}
+	encOut, err := openpgp.Encrypt(o, []*openpgp.Entity{e}, nil, h, nil)
```

Comment:
Maybe I'm missing something, but I think `[]*openpgp.Entity{e}` allows you to specify multiple recipients. You would have to change `Options.EncryptionKey` to a `[]string` and the CLI argument to a `StringSliceVar`, but I think we could support multiple public keys this way?

Reply:
That would make a lot of sense, but from my 🔍 the entities would need to be read using `openpgp.ReadKeyRing` and unfortunately that seems to be [broken](https://github.com/golang/go/issues/29082#issuecomment-809621751) 😞

Reply:
I don't think I follow. When would `openpgp.ReadKeyRing` have be used?

The way I'm imagining this would be that the user invokes this via `cilium sysdump --encryption-key default  --encryption-key my-own-key.gpg --encryption-key my-bosses-key.gpg`

This then would cause 3 invocations to `readDestinationEntity` - the default key would be read from the goembeded `pubKey`, the other two via `opengpg.ReadEntity`. All three are then appended to a `[]*openpgp.Entity` and passed here.

When decrypting, we will only use a single key anyways. I for example would use `my-own-key.gpg`.

Reply:
The following patch works for me, i.e. I can decrypt the sysdump via `gpg --decrypt` without the default private key installed:

```diff
diff --git a/internal/cli/cmd/sysdump.go b/internal/cli/cmd/sysdump.go
index a49923f51bf1..2083ce4d918d 100644
--- a/internal/cli/cmd/sysdump.go
+++ b/internal/cli/cmd/sysdump.go
@@ -106,8 +106,8 @@ func newCmdSysdump() *cobra.Command {
        cmd.Flags().BoolVar(&sysdumpOptions.Encrypt,
                "encrypt", sysdump.DefaultEncrypt,
                "Whether to encrypt the resulting zip file")
-       cmd.Flags().StringVar(&sysdumpOptions.EncryptionKey,
-               "encryption-key", sysdump.DefaultEncryptionKey,
+       cmd.Flags().StringSliceVar(&sysdumpOptions.EncryptionKeys,
+               "encryption-key", []string{sysdump.DefaultEncryptionKey},
                "The path to the OpenPGP-compliant public key to use for encrypting the resulting zip file. Defaults to an embeded OpenPGP-compliant public key")
 
        return cmd
diff --git a/sysdump/defaults.go b/sysdump/defaults.go
index d151bda62f7d..e6e8c0618da1 100644
--- a/sysdump/defaults.go
+++ b/sysdump/defaults.go
@@ -31,7 +31,7 @@ const (
        DefaultCiliumOperatorNamespace     = DefaultCiliumNamespace
        DefaultDebug                       = false
        DefaultEncrypt                     = true
-       DefaultEncryptionKey               = ""
+       DefaultEncryptionKey               = "default"
        DefaultHubbleLabelSelector         = labelPrefix + "hubble"
        DefaultHubbleNamespace             = DefaultCiliumNamespace
        DefaultHubbleRelayLabelSelector    = labelPrefix + "hubble-relay"
diff --git a/sysdump/encryption.go b/sysdump/encryption.go
index 44f2f7b42d24..3cc9f8552649 100644
--- a/sysdump/encryption.go
diff --git a/internal/cli/cmd/sysdump.go b/internal/cli/cmd/sysdump.go
index a49923f51bf1..2083ce4d918d 100644
--- a/internal/cli/cmd/sysdump.go
+++ b/internal/cli/cmd/sysdump.go
@@ -106,8 +106,8 @@ func newCmdSysdump() *cobra.Command {
        cmd.Flags().BoolVar(&sysdumpOptions.Encrypt,
                "encrypt", sysdump.DefaultEncrypt,
                "Whether to encrypt the resulting zip file")
-       cmd.Flags().StringVar(&sysdumpOptions.EncryptionKey,
-               "encryption-key", sysdump.DefaultEncryptionKey,
+       cmd.Flags().StringSliceVar(&sysdumpOptions.EncryptionKeys,
+               "encryption-key", []string{sysdump.DefaultEncryptionKey},
                "The path to the OpenPGP-compliant public key to use for encrypting the resulting zip file. Defaults to an embeded OpenPGP-compliant public key")
 
        return cmd
diff --git a/sysdump/defaults.go b/sysdump/defaults.go
index d151bda62f7d..e6e8c0618da1 100644
--- a/sysdump/defaults.go
+++ b/sysdump/defaults.go
@@ -31,7 +31,7 @@ const (
        DefaultCiliumOperatorNamespace     = DefaultCiliumNamespace
        DefaultDebug                       = false
        DefaultEncrypt                     = true
-       DefaultEncryptionKey               = ""
+       DefaultEncryptionKey               = "default"
        DefaultHubbleLabelSelector         = labelPrefix + "hubble"
        DefaultHubbleNamespace             = DefaultCiliumNamespace
        DefaultHubbleRelayLabelSelector    = labelPrefix + "hubble-relay"
diff --git a/sysdump/encryption.go b/sysdump/encryption.go
index 44f2f7b42d24..3cc9f8552649 100644
--- a/sysdump/encryption.go
+++ b/sysdump/encryption.go
@@ -29,11 +29,15 @@ import (
 //go:embed pubkey.gpg
 var pubKey []byte
 
-func createEncryptedZipFile(pathToZipFile string, pathToEncryptionKey string) (string, error) {
+func createEncryptedZipFile(pathToZipFile string, pathsToEncryptionKey []string) (string, error) {
        // Read the destination entity.
-       e, err := readDestinationEntity(pathToEncryptionKey)
-       if err != nil {
-               return "", err
+       var entityList []*openpgp.Entity
+       for _, k := range pathsToEncryptionKey {
+               e, err := readDestinationEntity(k)
+               if err != nil {
+                       return "", err
+               }
+               entityList = append(entityList, e)
        }
        // Create the destination file.
        d := pathToZipFile + ".gpg"
```

Reply:
@gandro oh I see now what you mean — read from several different public key files instead of from a keyring file containing the three public keys. I think that's very nice — do you want to push this patch to my branch? 🙂

Reply:
Done!